packages/synapse-react-client/src/components/doi/CreateOrUpdateDoiModal.tsx

```diff
@@ -0,0 +1,324 @@
+import { DialogBase } from '@/components/DialogBase'
+import { doiFormSchema, doiFormUiSchema } from '@/components/doi/DoiFormSchemas'
+import { JsonSchemaForm } from '@/components/JsonSchemaForm/JsonSchemaForm'
+import { MarkdownPopover } from '@/components/Markdown/MarkdownPopover'
+import { StyledFormControl } from '@/components/styled/StyledFormControl'
+import { displayToast } from '@/components/ToastMessage/ToastMessage'
+import { useCreateOrUpdateDOI, useGetDOI } from '@/synapse-queries/doi/useDOI'
+import {
+  useGetEntity,
+  useGetVersions,
+} from '@/synapse-queries/entity/useEntity'
+import { useGetCurrentUserProfile } from '@/synapse-queries/user/useUserBundle'
+import {
+  convertToEntityType,
+  entityTypeToFriendlyName,
+  isTable,
+  isVersionableEntity,
+} from '@/utils/functions/EntityTypeUtils'
+import { HelpTwoTone } from '@mui/icons-material'
+import {
+  Alert,
+  Button,
+  InputLabel,
+  MenuItem,
+  Select,
+  Typography,
+} from '@mui/material'
+import { DoiResourceTypeGeneral, V2Doi } from '@sage-bionetworks/synapse-client'
+import { EntityType } from '@sage-bionetworks/synapse-types'
+import { isEmpty } from 'lodash-es'
+import { useEffect, useState } from 'react'
+
+type CreateOrUpdateDoiModalProps = {
+  /** Whether the dialog is open */
+  open: boolean
+  /** Callback to call when the dialog is closed */
+  onClose: () => void
+  /** The type of object */
+  objectType: V2Doi['objectType']
+  /** The ID of the object */
+  objectId: string
+  /** The optional version number of the object used to populate the form. */
+  defaultVersionNumber?: number
+}
+
+type DoiFormData = {
+  titles?: string[]
+  creators?: string[]
+  resourceTypeGeneral?: string
+  publicationYear?: number
+}
+
+function getSuggestedResourceTypeGeneral(
+  type: EntityType,
+): DoiResourceTypeGeneral {
+  if (type === EntityType.PROJECT || type === EntityType.FOLDER) {
+    return DoiResourceTypeGeneral.Collection
+  } else {
+    return DoiResourceTypeGeneral.Dataset
+  }
+}
+
+function convertDoiToFormData(doi: V2Doi): DoiFormData {
+  return {
+    titles: (doi.titles || []).map(title => title.title!),
+    creators: (doi.creators || []).map(creator => creator.creatorName!),
+    resourceTypeGeneral: doi.resourceType?.resourceTypeGeneral,
+    publicationYear: doi.publicationYear,
+  }
+}
+
+/**
+ * Returns true if the DOI has fields that are not exposed in the UI, so we can let the user know that any changes
+ * may be destructive to hidden fields.
+ */
+function getWasModifiedViaAPI(doi: V2Doi) {
+  if (doi.creators?.some(creator => !isEmpty(creator.nameIdentifiers))) {
+    return true
+  }
+  return false
+}
```

Comment:
e.g. you can add an ORCiD to your DOI programmatically, but we don't show that in the UI